ASAN_TRAP | WebCore::Editor

roberto-apple opened the pull request:
#### 7bf00c604558a72ddbe23d64938b25d320e9d921
<pre>
ASAN_TRAP | WebCore::Editor
<a href="https://bugs.webkit.org/show_bug.cgi?id=290816">https://bugs.webkit.org/show_bug.cgi?id=290816</a>
<a href="https://rdar.apple.com/147936416">rdar://147936416</a>

Reviewed by Ryosuke Niwa.

Allow renderer to be any descendent of RenderBlock.

* LayoutTests/editing/editability/contenteditable-containing-block-no-crash-expected.txt: Added.
* LayoutTests/editing/editability/contenteditable-containing-block-no-crash.html: Added.
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::baseWritingDirectionForSelectionStart const):

Canonical link: <a href="https://commits.webkit.org/293191@main">https://commits.webkit.org/293191@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/beddff3b7f3c62a9f829da0675168fdd370ef880

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98157 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17788 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8015 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103274 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/48686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100202 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18080 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26239 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/74735 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/48686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101161 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/13702 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88688 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55095 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13486 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6625 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48128 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6704 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105650 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25243 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83719 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25616 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84871 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83174 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21010 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27815 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5498 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/18883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25202 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/30376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25022 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28338 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26597 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

roberto-apple commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2183,7 +2183,7 @@ WritingDirection Editor::baseWritingDirectionForSelectionStart() const
     if (renderer && !is<RenderBlockFlow>(*renderer))
         renderer = renderer->containingBlock();
-    if (!renderer)
+    if (!renderer || !is<RenderBlockFlow>(*renderer))
```
If execution hits `renderer = renderer->containingBlock();` and the newly assigned renderer is not of type RenderBlockFlow, the downcast below will hit a release assert.

alanbaradlay replied:
@rniwa do you know why we need to have a `RenderBlockFlow` here and not just a `RenderBlock`? I've never seen this function but judging by its name (`baseWritingDirectionForSelectionStart`) a `RenderBlock` may just be sufficient here?

rniwa replied:
yeah, `RenderBlock` should be fine.

roberto-apple replied:
Will update accordingly to use RenderBlock

rniwa commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2181,16 +2181,18 @@ WritingDirection Editor::baseWritingDirectionForSelectionStart() const
-        return WritingDirection::LeftToRight;
-    case TextDirection::RTL:
-        return WritingDirection::RightToLeft;
+    if (CheckedPtr renderBlock = dynamicDowncast<RenderBlock>(*renderer)) {
```
I don't think this will ever fail now that we're checking `is<RenderBlock>`. `containingBlock` above will always return `RenderBlock`.

roberto-apple replied:
Will update.

roberto-apple committed:
ASAN_TRAP | WebCore::Editor
https://bugs.webkit.org/show_bug.cgi?id=290816
rdar://147936416

Reviewed by Ryosuke Niwa.

Allow renderer to be any descendent of RenderBlock.

* LayoutTests/editing/editability/contenteditable-containing-block-no-crash-expected.txt: Added.
* LayoutTests/editing/editability/contenteditable-containing-block-no-crash.html: Added.
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::baseWritingDirectionForSelectionStart const):

Canonical link: https://commits.webkit.org/293191@main
